### PR TITLE
Issue #1317 Add uninstall command for add-ons

### PR DESCRIPTION
**Pull request by praveenkumar**
TODO
=====

- Test cases
- Documentation.

### CODE REVIEW

File: cmd/minishift/cmd/addon/addon_uninstall.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	"fmt"
+
+	"github.com/minishift/minishift/pkg/minishift/addon/manager"
+	"github.com/minishift/minishift/pkg/util/os/atexit"
+	"github.com/spf13/cobra"
+)
+
+var addonsUnInstallCmd = &cobra.Command{
+	Use:   "uninstall ADDON_NAME",
+	Short: "UnInstalls the specified add-on.",
```
Comment (hferentschik) [inline cmd/minishift/cmd/addon/addon_uninstall.go]: just 'uninstall', no capital 'I'

File: cmd/minishift/cmd/addon/addon_uninstall.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+	if !addOnManager.IsInstalled(addonName) {
+		atexit.ExitWithMessage(0, fmt.Sprintf(noAddOnMessage, addonName))
+	}
+
+	unInstallAddon(addOnManager, addonName)
+}
+
+func unInstallAddon(addOnManager *manager.AddOnManager, addOnName string) {
+	err := addOnManager.UnInstall(addOnName)
```
Comment (hferentschik) [inline cmd/minishift/cmd/addon/addon_uninstall.go]: _Uninstall_? Personally I see uninstall as a single word. Up to you on a method level. For user facing messages it is for sure 'uninstall'

File: pkg/minishift/addon/manager/addon_manager.go [hunk 3]
```diff
@@ -132,6 +132,23 @@ func (m *AddOnManager) Enable(addonName string, priority int) (*addon.AddOnConfi
 	return &addon.AddOnConfig{addonName, true, float64(priority)}, nil
 }
 
+// UnInstall uninstalls the addon specified via addonName.
+func (m *AddOnManager) UnInstall(addonName string) error {
```
Comment (hferentschik) [inline pkg/minishift/addon/manager/addon_manager.go]: :+1

**Comment by hferentschik** [comment]
Looking good. Tests and docs and we are talking ;-)

**Comment by praveenkumar** [comment]
@hferentschik Everything is in place now, have a look.

**Comment by hferentschik** [comment]
@praveenkumar Seems _Test_uninstall_with_enable_flag_works_ fails on Windows.

**Comment by hferentschik** [comment]
@praveenkumar not quite sure why CentOS CI failed. The errors seem to indicate that link_check got run without running synopsis_docs first. No idea how this would be possible.

**Comment by hferentschik** [comment]
retest this please

**Comment by praveenkumar** [comment]
> Seems Test_uninstall_with_enable_flag_works fails on Windows.

Right, I am looking into it, any idea why centos CI failed for doc lint check?

File: cmd/minishift/cmd/addon/addon_uninstall.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	if err != nil {
+		atexit.ExitWithMessage(1, fmt.Sprintf("Cannot Uninstall the add-on %s: %s", addOnName, err.Error()))
+	} else {
+		fmt.Println(fmt.Sprintf("Add-on '%s' Uninstalled", addOnName))
+	}
+
+	addOnConfigMap := GetAddOnConfiguration()
```
Comment (hferentschik) [inline cmd/minishift/cmd/addon/addon_uninstall.go]: What's about moving these three lines into a function in util.go. something like RemoveAddOnFromConfig(addOnName string)?

File: cmd/minishift/cmd/addon/addon_uninstall_test.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	tmpMinishiftHomeDir := cli.SetupTmpMinishiftHome(t)
+	tee := cli.CreateTee(t, true)
+	defer cli.TearDown(tmpMinishiftHomeDir, tee)
+
+	atexit.RegisterExitHandler(cli.VerifyExitCodeAndMessage(t, tee, 1, emptyAddOnError))
+
+	runUnInstallAddon(nil, nil)
+}
+
+func Test_uninstall_with_enable_flag_works(t *testing.T) {
```
Comment (hferentschik) [inline cmd/minishift/cmd/addon/addon_uninstall_test.go]: nice

File: cmd/minishift/cmd/addon/addon_uninstall_test.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+)
+
+var adminUser string = `# Name: admin-user
+# Description: Create admin user and assign the cluster-admin role to it.
+
+oc create user admin --full-name=admin
+oc adm policy add-cluster-role-to-user cluster-admin admin
+`
+
+func Test_addon_name_must_be_specified_for_uninstall_command(t *testing.T) {
```
Comment (hferentschik) [inline cmd/minishift/cmd/addon/addon_uninstall_test.go]: nice

File: docs/source/using/addons.adoc [hunk 7]
```diff
@@ -253,7 +253,7 @@ $ minishift addons apply anyuid admin-user
 
 Add-ons can be removed with the xref:../command-ref/minishift_addons_remove.adoc#[`minishift addons remove`] command.
 It is the mirror command to xref:../using/addons.adoc#apply-addons[`minishift addons apply`] command and similarly can be used regardless whether the add-on is enabled or not.
-Provided it exists, `minishift addons apply` will execute the commands specified in the file `<addon_name>.addon.remove`.
+Provided it exists, `minishift addons remove` will execute the commands specified in the file `<addon_name>.addon.remove`.
```
Comment (hferentschik) [inline docs/source/using/addons.adoc]: ahh, good catch

File: docs/source/using/addons.adoc [hunk 8]
```diff
@@ -267,6 +267,23 @@ $ minishift addons remove admin-user
 admin user deleted
 ----
 
+[[uninstall-addons]]
+== Uninstalling Add-ons
+
+Add-ons can be uninstall with the xref:../command-ref/minishift_addons_uninstall.adoc#[`minishift addons uninstall`] command.
+It is the mirror command to xref:../command-ref/minishift_addons_install.adoc#[`minishift addons install`] command and can be used regardless whether the add-on is enabled or not.
+Provided it exists, `minishift addons uninstall` will delete add-on specific directory.
```
Comment (hferentschik) [inline docs/source/using/addons.adoc]: "Provided it exists, `minishift addons uninstall` will delete the whole add-on directory."

**Comment by praveenkumar** [comment]
@hferentschik All suggestions are now taken care. Have an another look.

**Comment by hferentschik** [comment]
> All suggestions are now taken care. Have an another look.

Did you find the file handle culprit?

File: cmd/minishift/cmd/addon/util.go [hunk 9]
```diff
@@ -86,3 +86,10 @@ func fillStruct(data map[string]interface{}, result interface{}) {
 		val.Set(reflect.ValueOf(v))
 	}
 }
+
+// RemoveAddOnFromConfig remove given add-on name entry from config file
+func RemoveAddOnFromConfig(addOnName string) {
```
Comment (hferentschik) [inline cmd/minishift/cmd/addon/util.go]: +1

**Comment by praveenkumar** [comment]
> Did you find the file handle culprit?

@hferentschik yes looks like we were opening file in addon_parser for Reader but not closing it. I took another approach where we don't need to open a file instead just read it and store as string and then create reader using `strings` package. It is part of same PR.

File: pkg/minishift/addon/parser/addon_parser.go [hunk 10]
```diff
@@ -142,12 +141,12 @@ func (parser *AddOnParser) getAddOnContentReader(addOnDir string, fileSuffix str
...
 	}
 
-	file, err := os.Open(filepath.Join(addOnDir, addOnFiles[0]))
+	file, err := ioutil.ReadFile(filepath.Join(addOnDir, addOnFiles[0]))
 	if err != nil {
 		return nil, NewParseError(fmt.Sprintf("Unable to open addon definition '%s'", addOnFiles[0]), addOnDir, "")
 	}
-
-	return bufio.NewReader(file), nil
+	reader := strings.NewReader(string(file))
```
Comment (hferentschik) [inline pkg/minishift/addon/parser/addon_parser.go]: ahhh, here we go.

**Comment by hferentschik** [comment]
Merged, thanks. (changed docs a bit)

**Pull request closed without merging**